pkg/local_object_storage/pilorama/boltdb.go

```diff
@@ -23,6 +23,13 @@ type boltForest struct {
 
 	modeMtx sync.Mutex
 	mode    mode.Mode
+
+	// mtx protects batches field.
+	mtx     sync.Mutex
+	batches []*batch
```

Comment:
Do you mean why we use custom batches instead of `Batch`?

Inserting operations in a log is costly: we need to rollback and reapply some operations. If we insert 2 operations, with this PR we need to rollback+reapply only once, because of the custom logic.

Reply:
I'm rather talking about `DB.Begin()` and managing DB transactions manually. It'd be much simpler, but of course the result depends on how effective your time sorting in `(*batch).run` is. My expectation would be that it's not critical (so that most of the time operations are already well-ordered), but this entirely depends on the operations flow.

Reply:
The operation flow is comprised from 2 components:
1. Background sync.
2. Replication of user operations.

While operations from (2) should be near the end of the log, under high load the order of the last e.g. 100 operations is completely random.

Could you elaborate about `DB.Begin()`? How does it solve a problem?

Reply:
Whether it solves the problem or not depends on the core issue. If it's DB sync (which would be my primary suspect), then `DB.Begin()` can help a lot, because you'll open up a transaction, push a set of changes into it and then commit (with the same logic, either `MaxBatchSize` or `MaxBatchDelay`). If it's rollbacks, then sure, transactions may be of little help.

Reply:
The issue is that inserting N operation in random order has O(N^2) complexity.
This commit ammortizes the cost, so that if N operations are grouped in a single batch we have O(N) complexity.
For `append` I think native `bbolt.Batch` already does the job.

Basically, we block for too much time (because of multiple ineficcient undo+do in a row).
Sync can also be an issue, but it is not obvious how to approach it (we already use native `bbolt.Batch` everywhere).

Reply:
> For append I think native bbolt.Batch already does the job.

I wouldn't be so sure.

> Concurrent Batch calls are opportunistically combined into larger
transactions. Batch is only useful when there are multiple goroutines
calling it.

In fact, if you're doing all changes in a single thread it slows you down.

Reply:
We do it in different goroutines (each tree service RPC is executed in parallel)